lib/Runtime/Library/JavascriptProxy.cpp

```diff
@@ -919,11 +912,38 @@ namespace Js
         Var trapResult = getEnumeratorMethod->CallFunction(arguments);
         threadContext->SetImplicitCallFlags((Js::ImplicitCallFlags)(saveImplicitCallFlags | ImplicitCall_Accessor));
 
-        if (!JavascriptOperators::IsObject(trapResult))
+        if (!JavascriptOperators::IsArray(trapResult))
         {
-            JavascriptError::ThrowTypeError(scriptContext, JSERR_InconsistentTrapResult, _u("enumerate"));
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_InconsistentTrapResult, _u("ownKeys"));
         }
-        *enumerator = IteratorObjectEnumerator::Create(scriptContext, trapResult);
+
+        JavascriptArray* arrTrapResult = JavascriptArray::FromVar(trapResult);        
+        uint32 len = arrTrapResult->GetLength();
+        Var undefined = scriptContext->GetLibrary()->GetUndefined();
+        JsUtil::BaseDictionary<const char16*, Var, Recycler> dict(scriptContext->GetRecycler());
+        JavascriptArray* arrResult = scriptContext->GetLibrary()->CreateArray();
+        int index = 0;
+        arrTrapResult->ForEachItemInRange<false>(0, len, undefined, scriptContext, [&](uint32, Var element)
+        {
+            if (JavascriptString::Is(element))
+            {
+                Js::PropertyDescriptor desc;
+                JavascriptString* str = JavascriptString::FromVar(element);
+                BOOL ret = JavascriptOperators::GetOwnPropertyDescriptor(this, str, scriptContext, &desc);
+                if (ret && !dict.ContainsKey(str->GetSz()))
+                {
+                    dict.Add(str->GetSz(), element);
+                    if (desc.IsEnumerable()) 
+                    {
+                        arrResult->SetItem(index++, element, PropertyOperation_None);
+                    }
+                }
+            }
+        });
+
+        Var arrayIterator = JavascriptOperators::GetIterator(RecyclableObject::FromVar(arrResult), scriptContext);
```

Comment:
eventually we returns an enumerator. Why do we need to create an enumerator over the iterator? can we just create an enumerator off the array directly?